File: webapp/src/model/volunteer.js
import User from './user';
import Roles from './role';
import Status from './status';
import Company from './company';

export default class{
    constructor(user = new User(), company = new Company(), schooling = '',
        conclusion = false, course ='', rg = '', answers = [], ratings = []) {
        user.role = Roles.VOLUNTEER;
        user.status = Status.REGISTER;
        this.user = user;
        this.company = company;
        this.schooling = schooling;
        this.conclusion = conclusion;
        this.course = course;
        this.rg = rg;
        this.answers = answers;
        this.ratings = ratings;
    }
}

File: src/main/java/br/ufsc/cultivar/model/Volunteer.java
package br.ufsc.cultivar.model;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Value;
import lombok.experimental.FieldDefaults;
import lombok.experimental.Wither;

import javax.validation.Valid;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Value
@Wither
@Builder(builderClassName = "Builder")
@JsonDeserialize(builder = Volunteer.Builder.class)
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class Volunteer {
    @Valid
    @NotNull
    User user;
    @Valid
    @NotNull
    Company company;
    @NotNull
    Schooling schooling;
    @NotNull
    Boolean conclusion;
    @NotBlank
    @NotNull
    String course;
    @NotBlank
    @NotNull
    String rg;
    School school;
    List<@Valid Dispatch> dispatches;
    @NotNull
    @NotEmpty
    List<Answer> answers;
    List<@Valid Rating> ratings;

    public Double getRating() {
        return Optional.ofNullable(ratings)
                .orElse(new ArrayList<>())
                .stream()
                .mapToInt(Rating::getGrade)
                .average()
                .orElse(0.);
    }

    @JsonPOJOBuilder(withPrefix = "")
    public static class Builder{
        public Builder(){}
    }
}
